<center> <h1>Homework 1</h1> </center>

<center> <t>Authors: EL Fattahi Charaf-Ed-Dine</t> </center>
<center> <t>         Rupesh Kumar</t> </center>

# Exercise 1

Consider the second stage problem defined by
\begin{align*}
    \min_y\ & 2y_1 +y_2 \\
    \mbox{s.t. } & y_1+y_2 \leq 1-x_1,\\
                 & y_1\leq\xi-x_1-x_2\\
                 & y_1,y_2 \geq 0
\end{align*}
Show that this program has complete recourse if $\mathbb{E}\left[\xi\right]$ is finite.

### answer

We start by rewriting the constraints in terms of equalities:
\begin{align*}
    & y_1 + y_2 - y_3 = 1 - x_1,\\
    & y_1 - y_4 = \xi - x_1 - x_2\\
    & y_1,y_2,y_3,y_4 \geq 0
\end{align*}
So
\begin{align*}
    W = \begin{bmatrix}
            1 & 1 & -1 & 0\\
            1 & 0 & 0 & -1
        \end{bmatrix}
\end{align*}
To show that the recourse is complete we need to verify the following:
\begin{align*}
    \forall z\in\mathcal{R}^2,\ \exists y\in\mathcal{R_+^4} \text{   s.t.   } Wy=z 
\end{align*}
Let $z\in\mathcal{R}^2$. 

We put $y_1=\max\left\{0,z_1,z_2\right\}$, $y_2=0$, $y_3=y_1-z_1$ and $y_4=y_1-z_2$. This choice clearly satisfies $Wy=z$ and $y\geq0$. Therefore the problem has complete recourse.

$\mathbb{E}\left[\xi\right]$ being finite is irrelevant in the proof.

# Exercise 2

Consider the second stage program
\begin{align*}
    \min_y\ & 2y_1 +y_2 \\
    \mbox{s.t. } & y_1-y_2 \leq 2-\xi x_1,\\
                 & y_2\leq x_2\\
                 & y_1,y_2 \geq 0
\end{align*}
Find $K_2(\xi)$ and $K_2$ for:

$\textbf{(a)}$ $\xi\sim U[0,1]$

\begin{align*}
    K_2 &= \underset{\xi\in[0,1]}{\bigcap} K_2(\xi)\\
    &= \underset{\xi\in[0,1]}{\bigcap} \left\{x:\ x_1\leq\frac{2+x_2}{\xi},\ x_2\geq0\right\}\\
    &= \left\{x:\ x_1-x_2\leq 2,\ x_2\geq0\right\}.
\end{align*}

$\textbf{(b)}$ $\xi\sim$Poisson$(\lambda)$, $\lambda>0$.

\begin{align*}
    K_2 &= \underset{\xi}{\bigcap} K_2(\xi)\\
    &= \underset{\xi}{\bigcap} \left\{x:\ x_1\leq\frac{2+x_2}{\xi},\ x_2\geq0\right\}\\
    &= \left\{x:\ x_1\leq 0,\ x_2\geq0\right\},
\end{align*}
since $K_2(\xi)$ restricts $x_1$ more and more as $\xi$ grows.

# Exercise 3

We consider a baker who decides his production in the morning and can sell his production during the day at a profit. In the evening, the surplus can be sold at a reduced price as the bread has lost its freshness. Two types of bread can be produced: white bread and whole wheat bread. White bread has a production cost of 1.5$\$$ each, and whole wheat bread is 1.8$\$$. They are priced at 3$\$$ and 4$\$$ each, respectively. The unsold products are sold in the evening at 1$\$$ and 1.2$\$$. The production of black bread requires 200g of flour, while for white bread it takes 150g. The baker has 12kg of flour at the start of the day. Assume that the demand for white bread follows a normal of mean 50 and standard deviation of 5, and that the demand for whole wheat bread follows a normal of mean 30 and standard deviation of 2. De plus, the two demands are correlated, and the covariance between the two demands is 0.4.

We aim to minimize the total cost (investment and production costs).
Formulate the problem and solve it using the L-Shaped method.
You have to provide the code for L-Shaped implementation and properly comment it.

## Question 1
Describe the problem as a two-step stochastic problem. Explain the choice of the first and second step decision variables and formulate the mathematical problem.

In the morning, $x_1$ white breads and $x_2$ whole wheat bread is produced. The first step decision variables are:

$x_1$ : Number of White bread to be produced in advanced

$x_2$ : Number of Whole White bread to be produced in advanced

The second step decision variables are:

$y_1$: effective sales of white bread in the morning

$y_2$: effective sales of whole wheat bread in the morning

$w_1$: number of unsold white bread

$w_2$: number of unsold whole wheat bread

The demand for white bread and whole wheat bread varies over days and is described by a random
variable $\xi$ (where $\xi_1\sim \mathcal{N}(50,5^2)$ for white bread and $\xi_2\sim \mathcal{N}(30,2^2)$ for whole wheat bread ).

The two demands are correlated, $\text{cov}(\xi_1,\xi_2)=0.4$.

We denote by $c_i$ the cost of production for each type of bread ($c_1=1.5$ for the white bread and $c_2=1.8$ for the whole wheat bread). Let $q_i$ be the selling price in the morning ($q_1=3$ for the white bread and $q_2=4$ for the whole wheat bread) and $r_i$ the selling price at the end of the day ($r_1=1$ for the white bread and $r_2=1.2$ for the whole wheat bread).

The problem is
\begin{align*}
    \min_x & \sum_{i=1}^2 c_i x_i + E_ξ[Q(x,ξ)] \\
    \mbox{s.t. } & 150 x_1+ 200 x_2 \leq 12000 \\
                 & x \geq 0
\end{align*}
where
\begin{align*} 
    Q(x,ξ) = \min_y\ & -\sum_{i=1}^2  q_{i} y_{i} + r_iw_i \\
    \mbox{s.t. } & y \leq \xi,\\
                 & y+w\leq x\\
                 & y=\min(\xi,x)\\
                 & w=\max(0, x-\xi)\\
                 & y \geq 0, \quad w\geq0
\end{align*}


## Question 2
Explain $K_1$, $K_2$ and $K_2^P$. Is the recourse complete, relatively complete or simple?

We have
\begin{align*}
    K_1 &= \left\{ x | 150 x_1 + 200 x_2 \leq 12000,\ x\geq0 \right\} \\
        &= \left\{ x | 3 x_1 + 4 x_2 \leq 240,\ x\geq0 \right\} \\
\end{align*}

For the second stage feasability

\begin{align*}
    K_2\left(\xi\right) &= \left\{ x | x\geq0 \right\} \\
\end{align*}

\begin{align*}
    K_2^P &= \cap_{\xi} K_2\left(\xi\right) \\
          &= \left\{ x | x\geq0 \right\}
\end{align*}

We have $K_1\subset K_2$. Therefore the recourse is relatively complete.

## Question 3
Sample 100 demand scenarios and implement the problem using Julia and JuMP.

In [1]:
using JuMP
using LinearAlgebra
using Statistics

In [2]:
using SparseArrays  # allows to benefit from sparse matrices

In [3]:
# import Pkg; Pkg.add("MathOptInterface")
MOI.INFEASIBLE

INFEASIBLE::TerminationStatusCode = 2

In [4]:
# import Pkg; Pkg.add("Ipopt")

In [5]:
using Gurobi
# using GLPK
# using Ipopt

solver = Gurobi.Optimizer
# solver = GLPK.Optimizer
# solver = Ipopt.Optimizer

Gurobi.Optimizer

## Extended form

We start by considering the extensive form in order to have a reference solution.

In [6]:
# we define constant variables
productionCosts = [1.5, 1.8]
productionFlour = [150, 200]
q = [3, 4]
r = [1, 1.2]
maxFlour = 12000
nBreadTypes = 2

2

In [7]:
using Random, Distributions
Random.seed!(123) # Setting the seed

MersenneTwister(123)

In [8]:
# We do a Cholesky decomposition to generate correlated scenarios
nScenarios = 100

μ = [50, 30]
σ = [5, 2]
covar = 0.4
rho = covar/(σ[1]*σ[2])

println("Reqquired Correlation is ", rho)

z = [rand(Normal(0, 1), nScenarios), rand(Normal(0, 1), nScenarios)]

X = μ[1] .+ σ[1].*z[1]
Y = μ[2] .+ σ[2].*(z[1].*rho + z[2].*(1-rho^2)^0.5)

println(cor(X, Y)) # to get more accurate cor and cov, take more scenarios
println(cov(X, Y))

X = reshape(X,1,nScenarios)
Y = reshape(Y,1,nScenarios)
println(size(X))

Reqquired Correlation is 0.04
0.09468968411666775
1.066970115550481
(1, 100)


In [9]:
m = Model()

scenarios = [X; Y]
prob = 1/100 # probability does not matter here (assuming i.i.d. variables)

@variable(m, x[1:nBreadTypes] >= 0)
@variable(m, y[1:nScenarios, 1:nBreadTypes] >= 0)
@variable(m, w[1:nScenarios, 1:nBreadTypes] >= 0)

@constraint(m, sum(productionFlour[i]*x[i] for i in 1:nBreadTypes) <= maxFlour)

for s = 1:nScenarios
    for j = 1:nBreadTypes
        @constraint(m, y[s,j] <= scenarios[j, s])
        @constraint(m, y[s,j] + w[s,j] <= x[j])
        @constraint(m, w[s,j] >= x[j]-scenarios[j, s])
        @constraint(m, y[s,j] <= x[j])
    end
end

@objective(m, Min, sum(productionCosts[i]*x[i] for i in 1:nBreadTypes) -
    sum(prob*sum((q[j]*y[s, j] + r[j]*w[s, j]) for j = 1:nBreadTypes) for s=1:nScenarios)
)

println(m)

Min 1.5 x[1] + 1.8 x[2] - 0.03 y[1,1] - 0.01 w[1,1] - 0.04 y[1,2] - 0.012 w[1,2] - 0.03 y[2,1] - 0.01 w[2,1] - 0.04 y[2,2] - 0.012 w[2,2] - 0.03 y[3,1] - 0.01 w[3,1] - 0.04 y[3,2] - 0.012 w[3,2] - 0.03 y[4,1] - 0.01 w[4,1] - 0.04 y[4,2] - 0.012 w[4,2] - 0.03 y[5,1] - 0.01 w[5,1] - 0.04 y[5,2] - 0.012 w[5,2] - 0.03 y[6,1] - 0.01 w[6,1] - 0.04 y[6,2] - 0.012 w[6,2] - 0.03 y[7,1] - 0.01 w[7,1] - 0.04 y[7,2] - 0.012 w[7,2] - 0.03 y[8,1] - 0.01 w[8,1] - 0.04 y[8,2] - 0.012 w[8,2] - 0.03 y[9,1] - 0.01 w[9,1] - 0.04 y[9,2] - 0.012 w[9,2] - 0.03 y[10,1] - 0.01 w[10,1] - 0.04 y[10,2] - 0.012 w[10,2] - 0.03 y[11,1] - 0.01 w[11,1] - 0.04 y[11,2] - 0.012 w[11,2] - 0.03 y[12,1] - 0.01 w[12,1] - 0.04 y[12,2] - 0.012 w[12,2] - 0.03 y[13,1] - 0.01 w[13,1] - 0.04 y[13,2] - 0.012 w[13,2] - 0.03 y[14,1] - 0.01 w[14,1] - 0.04 y[14,2] - 0.012 w[14,2] - 0.03 y[15,1] - 0.01 w[15,1] - 0.04 y[15,2] - 0.012 w[15,2] - 0.03 y[16,1] - 0.01 w[16,1] - 0.04 y[16,2] - 0.012 w[16,2] - 0.03 y[17,1] - 0.01 w[17,1] - 0.04

 y[13,1] ≤ 45.55532126551347
 -x[1] + y[13,1] + w[13,1] ≤ 0.0
 -x[1] + y[13,1] ≤ 0.0
 y[13,2] ≤ 29.482820827357322
 -x[2] + y[13,2] + w[13,2] ≤ 0.0
 -x[2] + y[13,2] ≤ 0.0
 y[14,1] ≤ 51.63607476935736
 -x[1] + y[14,1] + w[14,1] ≤ 0.0
 -x[1] + y[14,1] ≤ 0.0
 y[14,2] ≤ 29.146178113770233
 -x[2] + y[14,2] + w[14,2] ≤ 0.0
 -x[2] + y[14,2] ≤ 0.0
 y[15,1] ≤ 52.9620162133636
 -x[1] + y[15,1] + w[15,1] ≤ 0.0
 -x[1] + y[15,1] ≤ 0.0
 y[15,2] ≤ 26.506540502438888
 -x[2] + y[15,2] + w[15,2] ≤ 0.0
 -x[2] + y[15,2] ≤ 0.0
 y[16,1] ≤ 51.840010115322954
 -x[1] + y[16,1] + w[16,1] ≤ 0.0
 -x[1] + y[16,1] ≤ 0.0
 y[16,2] ≤ 31.3316286453197
 -x[2] + y[16,2] + w[16,2] ≤ 0.0
 -x[2] + y[16,2] ≤ 0.0
 y[17,1] ≤ 48.594333779470794
 -x[1] + y[17,1] + w[17,1] ≤ 0.0
 -x[1] + y[17,1] ≤ 0.0
 y[17,2] ≤ 30.77227199389245
 -x[2] + y[17,2] + w[17,2] ≤ 0.0
 -x[2] + y[17,2] ≤ 0.0
 y[18,1] ≤ 46.32557204967469
 -x[1] + y[18,1] + w[18,1] ≤ 0.0
 -x[1] + y[18,1] ≤ 0.0
 y[18,2] ≤ 30.3624695636054
 -x[2] + y[18,2] + w[18,2] ≤ 0.0
 

 -x[1] + y[69,1] ≤ 0.0
 y[69,2] ≤ 28.893203296967446
 -x[2] + y[69,2] + w[69,2] ≤ 0.0
 -x[2] + y[69,2] ≤ 0.0
 y[70,1] ≤ 63.867975226360144
 -x[1] + y[70,1] + w[70,1] ≤ 0.0
 -x[1] + y[70,1] ≤ 0.0
 y[70,2] ≤ 28.124900943647855
 -x[2] + y[70,2] + w[70,2] ≤ 0.0
 -x[2] + y[70,2] ≤ 0.0
 y[71,1] ≤ 44.832839779794696
 -x[1] + y[71,1] + w[71,1] ≤ 0.0
 -x[1] + y[71,1] ≤ 0.0
 y[71,2] ≤ 30.93319557520937
 -x[2] + y[71,2] + w[71,2] ≤ 0.0
 -x[2] + y[71,2] ≤ 0.0
 y[72,1] ≤ 48.47769656730959
 -x[1] + y[72,1] + w[72,1] ≤ 0.0
 -x[1] + y[72,1] ≤ 0.0
 y[72,2] ≤ 27.295742107169527
 -x[2] + y[72,2] + w[72,2] ≤ 0.0
 -x[2] + y[72,2] ≤ 0.0
 y[73,1] ≤ 54.24485998809533
 -x[1] + y[73,1] + w[73,1] ≤ 0.0
 -x[1] + y[73,1] ≤ 0.0
 y[73,2] ≤ 30.757265920672538
 -x[2] + y[73,2] + w[73,2] ≤ 0.0
 -x[2] + y[73,2] ≤ 0.0
 y[74,1] ≤ 35.42246970030651
 -x[1] + y[74,1] + w[74,1] ≤ 0.0
 -x[1] + y[74,1] ≤ 0.0
 y[74,2] ≤ 27.60485884925647
 -x[2] + y[74,2] + w[74,2] ≤ 0.0
 -x[2] + y[74,2] ≤ 0.0
 y[75,1] ≤ 49.9236070604481
 -x[1] +

 w[63,1] ≥ 0.0
 w[64,1] ≥ 0.0
 w[65,1] ≥ 0.0
 w[66,1] ≥ 0.0
 w[67,1] ≥ 0.0
 w[68,1] ≥ 0.0
 w[69,1] ≥ 0.0
 w[70,1] ≥ 0.0
 w[71,1] ≥ 0.0
 w[72,1] ≥ 0.0
 w[73,1] ≥ 0.0
 w[74,1] ≥ 0.0
 w[75,1] ≥ 0.0
 w[76,1] ≥ 0.0
 w[77,1] ≥ 0.0
 w[78,1] ≥ 0.0
 w[79,1] ≥ 0.0
 w[80,1] ≥ 0.0
 w[81,1] ≥ 0.0
 w[82,1] ≥ 0.0
 w[83,1] ≥ 0.0
 w[84,1] ≥ 0.0
 w[85,1] ≥ 0.0
 w[86,1] ≥ 0.0
 w[87,1] ≥ 0.0
 w[88,1] ≥ 0.0
 w[89,1] ≥ 0.0
 w[90,1] ≥ 0.0
 w[91,1] ≥ 0.0
 w[92,1] ≥ 0.0
 w[93,1] ≥ 0.0
 w[94,1] ≥ 0.0
 w[95,1] ≥ 0.0
 w[96,1] ≥ 0.0
 w[97,1] ≥ 0.0
 w[98,1] ≥ 0.0
 w[99,1] ≥ 0.0
 w[100,1] ≥ 0.0
 w[1,2] ≥ 0.0
 w[2,2] ≥ 0.0
 w[3,2] ≥ 0.0
 w[4,2] ≥ 0.0
 w[5,2] ≥ 0.0
 w[6,2] ≥ 0.0
 w[7,2] ≥ 0.0
 w[8,2] ≥ 0.0
 w[9,2] ≥ 0.0
 w[10,2] ≥ 0.0
 w[11,2] ≥ 0.0
 w[12,2] ≥ 0.0
 w[13,2] ≥ 0.0
 w[14,2] ≥ 0.0
 w[15,2] ≥ 0.0
 w[16,2] ≥ 0.0
 w[17,2] ≥ 0.0
 w[18,2] ≥ 0.0
 w[19,2] ≥ 0.0
 w[20,2] ≥ 0.0
 w[21,2] ≥ 0.0
 w[22,2] ≥ 0.0
 w[23,2] ≥ 0.0
 w[24,2] ≥ 0.0
 w[25,2] ≥ 0.0
 w[26,2] ≥ 0.0
 w[27,2] ≥ 0.0
 w[28,2] ≥ 0.0
 w[29,2] ≥ 0.0
 w[

In [10]:
set_optimizer(m, solver)
optimize!(m)

Academic license - for non-commercial use only - expires 2022-10-15
Gurobi Optimizer version 9.1.2 build v9.1.2rc0 (mac64)
Thread count: 4 physical cores, 8 logical processors, using up to 8 threads
Optimize a model with 801 rows, 402 columns and 1602 nonzeros
Model fingerprint: 0x05301ffc
Coefficient statistics:
  Matrix range     [1e+00, 2e+02]
  Objective range  [1e-02, 2e+00]
  Bounds range     [0e+00, 0e+00]
  RHS range        [3e+01, 1e+04]
Presolve removed 200 rows and 0 columns
Presolve time: 0.00s
Presolved: 601 rows, 402 columns, 1402 nonzeros

Iteration    Objective       Primal Inf.    Dual Inf.      Time
       0   -2.9475282e+02   1.890137e+03   0.000000e+00      0s
     315   -1.2466843e+02   0.000000e+00   0.000000e+00      0s

Solved in 315 iterations and 0.00 seconds
Optimal objective -1.246684338e+02

User-callback calls 346, time in user-callback 0.00 sec


In [11]:
value.(x)

2-element Vector{Float64}:
 43.03340426813959
 27.724946798895306

In [12]:
objective_value(m)

-124.66843383234793

The solution is therefore $x = (43.033,\ 27.725)$.

## 2-stage formulation

We first create the first-stage problem.

In [13]:
function firststage()
    m = Model(with_optimizer(solver))

    @variable(m, x[1:nBreadTypes] >= 0)
    @variable(m, θ)
    
    @constraint(m, sum(productionFlour[i]*x[i] for i in 1:nBreadTypes) <= maxFlour)
    
    @objective(m, Min, sum(productionCosts[i]*x[i] for i in 1:nBreadTypes))
    
    return m, x, θ
end

firststage (generic function with 1 method)

In [14]:
function master_objective(m::Model, x, θ)
    @objective(m, Min, sum(productionCosts[i]*x[i] for i in 1:nBreadTypes) + θ)
    return m
end

master_objective (generic function with 1 method)

We then create the second-stage problem.

In [15]:
function secondstageCore(x, ξ)
    m = Model()

    @variable(m, y[1:nBreadTypes] >= 0)
    @variable(m, w[1:nBreadTypes] >= 0)
    
    # We would gain to use sparse matrices
    T=[0 0; 0 0; -1 0; 0 -1; -1 0; 0 -1; -1 0; 0 -1]
    h = [ξ[1]; ξ[2]; 0; 0; 0; 0; -ξ[1]; -ξ[2]]
    
    recourseConstraints = []

    return m, y, w, recourseConstraints, h, T
end 

secondstageCore (generic function with 1 method)

In [16]:
function secondstage(x, ξ)
    m, y, w, recourseConstraints, h, T = secondstageCore(x, ξ)
    
    for i = 1:nBreadTypes
        push!(recourseConstraints, @constraint(m, y[i] <= ξ[i]))
    end
    
    for i = 1:nBreadTypes
        push!(recourseConstraints, @constraint(m, y[i] <= x[i]))
    end
    
    for i = 1:nBreadTypes
        push!(recourseConstraints, @constraint(m, y[i] + w[i] <= x[i]))
    end
    
    for i = 1:nBreadTypes
        push!(recourseConstraints, @constraint(m, w[i] >= x[i]-ξ[i]))
    end
    
    @objective(m, Min, -sum(q[i]*y[i]+r[i]*w[i] for i = 1:nBreadTypes))

    set_optimizer(m, solver)
    status = optimize!(m)
 
    return m, recourseConstraints, h, T
end

secondstage (generic function with 1 method)

In [17]:
function secondstagefeasibility(x, ξ)
    m, y, w, recourseConstraints, h, T = secondstageCore(x, ξ)

    t = 4*nBreadTypes
    
    @variable(m, v[1:t] >= 0)

    for i = 1:nBreadTypes
        push!(recourseConstraints, @constraint(m, y[i] - v[i] <= ξ[i]))
    end

    for i = 1:nBreadTypes
        push!(recourseConstraints, @constraint(m, y[i] - v[2+i] <= x[i]))
    end
    
    for i = 1:nBreadTypes
        push!(recourseConstraints, @constraint(m, y[i] + w[i] - v[4+i] <= x[i]))
    end

    for i = 1:nBreadTypes
        push!(recourseConstraints, @constraint(m, w[i] + v[6+i] >= -ξ[i] + x[i]))
    end

    @objective(m, Min, sum(v[i] for i in 1:t))

    set_optimizer(m, solver)
    optimize!(m)
    σ = dual.(recourseConstraints)  # since the objective function is a minimization, we have the correct sign.

    return σ, h, T
end

secondstagefeasibility (generic function with 1 method)

In [18]:
function stop(Q, θ, k)
    nmax = 200
    tol = 1e-10
    if ((θ >= Q-tol) || (k == nmax))
        return true
    else
        return false
    end
end

stop (generic function with 1 method)

In [19]:
function lshaped(scenarios, prob)
    nScenarios = length(scenarios[1])
    
    k = 0       # iteration index
    nfcuts = 0  # number of feasibility cuts
    nocuts = 0  # number of optimality cuts
    
    first, x, θ = firststage()
    n = length(x)

    Q = +Inf
    valθ = -Inf
    
    while (!stop(Q, valθ, k))
        k += 1

        println(first)

        optimize!(first)
        status = termination_status(first)
        
        if (status != MOI.OPTIMAL)
            println("Error: status ", status)
            return status, x, first
        end
        xsol = value.(x)

        E = zeros(n)'
        e = 0.0
        Q = 0.0
        
        # Solve the second-stage programs
        
        for i = 1:nScenarios
            # p = prob[1][i] * prob[2][i]
            p = 1
                
            scs = transpose(scenarios)
            
            m, scstrs, h, T = secondstage(xsol, scs[i,:])
            status = termination_status(m)
            if (status == MOI.INFEASIBLE)
                # Build a feasibility cut
                σ, h, T = secondstagefeasibility(xsol, scs[i])
                E = σ'*T
                @constraint(first, sum(E[i]*x[i] for i in 1:nBreadTypes) >= σ'*h)
                nfcuts += 1
                break;
            elseif (status == MOI.OPTIMAL)
                Q += p*objective_value(m)
                # Build the optimality cut component
                π = dual.(scstrs)
                E += p*(π'*T)
                e += p*(dot(π,h))
            else
                println("Error second-stage resolution: status ", status)
                return status, x, first
            end
        end
        
        if (status == MOI.OPTIMAL)
            # add an optimality cut if it improves the lower bound
            if (nocuts == 0)
                # add θ in the problem
                @constraint(first, con, sum(E[i]*x[i] for i in 1:nBreadTypes) + θ >= e)
                master_objective(first, x, θ)
            else
                valθ = value.(θ)
                if (valθ < Q)
                    @constraint(first, sum(E[i]*x[i] for i in 1:nBreadTypes) + θ >= e)
                end
            end
            nocuts += 1
        end
        println(e)
    end
    
    println("Solved in ", k, " iterations.")
    return x, first
end

lshaped (generic function with 1 method)

In [20]:
x, firstst = lshaped(scenarios, prob)

Academic license - for non-commercial use only - expires 2022-10-15
Min 1.5 x[1] + 1.8 x[2]
Subject to
 150 x[1] + 200 x[2] ≤ 12000.0
 x[1] ≥ 0.0
 x[2] ≥ 0.0

Gurobi Optimizer version 9.1.2 build v9.1.2rc0 (mac64)
Thread count: 4 physical cores, 8 logical processors, using up to 8 threads
Optimize a model with 1 rows, 3 columns and 2 nonzeros
Model fingerprint: 0xa6855b61
Coefficient statistics:
  Matrix range     [2e+02, 2e+02]
  Objective range  [2e+00, 2e+00]
  Bounds range     [0e+00, 0e+00]
  RHS range        [1e+04, 1e+04]
Presolve removed 1 rows and 3 columns
Presolve time: 0.00s
Presolve: All rows and columns removed
Iteration    Objective       Primal Inf.    Dual Inf.      Time
       0    0.0000000e+00   0.000000e+00   2.000000e-06      0s

Solved in 0 iterations and 0.00 seconds
Optimal objective  0.000000000e+00

User-callback calls 25, time in user-callback 0.00 sec
Academic license - for non-commercial use only - expires 2022-10-15
Gurobi Optimizer version 9.1.2 build v9

(VariableRef[x[1], x[2]], A JuMP Model
Minimization problem with:
Variables: 3
Objective function type: AffExpr
`AffExpr`-in-`MathOptInterface.GreaterThan{Float64}`: 2 constraints
`AffExpr`-in-`MathOptInterface.LessThan{Float64}`: 1 constraint
`VariableRef`-in-`MathOptInterface.GreaterThan{Float64}`: 2 constraints
Model mode: AUTOMATIC
CachingOptimizer state: ATTACHED_OPTIMIZER
Solver name: Gurobi
Names registered in the model: con, x, θ)

The resulting master program is:

In [21]:
firstst

A JuMP Model
Minimization problem with:
Variables: 3
Objective function type: AffExpr
`AffExpr`-in-`MathOptInterface.GreaterThan{Float64}`: 2 constraints
`AffExpr`-in-`MathOptInterface.LessThan{Float64}`: 1 constraint
`VariableRef`-in-`MathOptInterface.GreaterThan{Float64}`: 2 constraints
Model mode: AUTOMATIC
CachingOptimizer state: ATTACHED_OPTIMIZER
Solver name: Gurobi
Names registered in the model: con, x, θ

In [22]:
objective_value(firstst)

-126.16942951784517

In [23]:
value.(x)

2-element Vector{Float64}:
 38.8704698810321
 30.847147589225926

The solution is therefore $x = (38.870,\ 30.847)$.

## Question 4
Interpret the solution.

The l-shaped method solution is $x = (38.870,\ 30.847)$.

We see that $150x_1 +  200x_2 = 12000$. This makes sense as we have the expected demand saturates completely first stage constaraint, i.e.
\begin{equation}
    50 \cdot 150 + 30 \cdot 200 = 13500 > 12000.
\end{equation}
So it makes sense to use all the availible flour.

One other observation is that the produced quantity of whole wheat bread meets the expected value of the demand, i.e $x_2\sim30$ while the white bread production is ar away from its expected demand, i.e $x_1<50$. This is reasonable as the whole wheat bread makes more profit. The production of the white bread is then dictated by how much flour do we still have.

## Question 5
Adapt the Julia code to obtain integer solutions.

We just have to add an additional constraint:

`for j = 1:nBreadTypes
    @constraint(m, x[j] in MOI.Integer())
end`

In [24]:
function firststage()
    m = Model(with_optimizer(solver))

    @variable(m, x[1:nBreadTypes] >= 0)
    @variable(m, θ)
    
    @constraint(m, sum(productionFlour[i]*x[i] for i in 1:nBreadTypes) <= maxFlour)
    
    for j = 1:nBreadTypes
        @constraint(m, x[j] in MOI.Integer())
    end
    
    @objective(m, Min, sum(productionCosts[i]*x[i] for i in 1:nBreadTypes))
    
    return m, x, θ
end

firststage (generic function with 1 method)

In [25]:
x, firstst = lshaped(scenarios, prob)

Academic license - for non-commercial use only - expires 2022-10-15
Min 1.5 x[1] + 1.8 x[2]
Subject to
 150 x[1] + 200 x[2] ≤ 12000.0
 x[1] ≥ 0.0
 x[2] ≥ 0.0
 x[1] integer
 x[2] integer

Gurobi Optimizer version 9.1.2 build v9.1.2rc0 (mac64)
Thread count: 4 physical cores, 8 logical processors, using up to 8 threads
Optimize a model with 1 rows, 3 columns and 2 nonzeros
Model fingerprint: 0xe3ed7922
Variable types: 1 continuous, 2 integer (0 binary)
Coefficient statistics:
  Matrix range     [2e+02, 2e+02]
  Objective range  [2e+00, 2e+00]
  Bounds range     [0e+00, 0e+00]
  RHS range        [1e+04, 1e+04]
Found heuristic solution: objective 0.0000000

Explored 0 nodes (0 simplex iterations) in 0.00 seconds
Thread count was 1 (of 8 available processors)

Solution count 1: 0 

Optimal solution found (tolerance 1.00e-04)
Best objective 0.000000000000e+00, best bound 0.000000000000e+00, gap 0.0000%

User-callback calls 21, time in user-callback 0.00 sec
Academic license - for non-commerci

(VariableRef[x[1], x[2]], A JuMP Model
Minimization problem with:
Variables: 3
Objective function type: AffExpr
`AffExpr`-in-`MathOptInterface.GreaterThan{Float64}`: 2 constraints
`AffExpr`-in-`MathOptInterface.LessThan{Float64}`: 1 constraint
`VariableRef`-in-`MathOptInterface.GreaterThan{Float64}`: 2 constraints
`VariableRef`-in-`MathOptInterface.Integer`: 2 constraints
Model mode: AUTOMATIC
CachingOptimizer state: ATTACHED_OPTIMIZER
Solver name: Gurobi
Names registered in the model: con, x, θ)

In [26]:
firstst

A JuMP Model
Minimization problem with:
Variables: 3
Objective function type: AffExpr
`AffExpr`-in-`MathOptInterface.GreaterThan{Float64}`: 2 constraints
`AffExpr`-in-`MathOptInterface.LessThan{Float64}`: 1 constraint
`VariableRef`-in-`MathOptInterface.GreaterThan{Float64}`: 2 constraints
`VariableRef`-in-`MathOptInterface.Integer`: 2 constraints
Model mode: AUTOMATIC
CachingOptimizer state: ATTACHED_OPTIMIZER
Solver name: Gurobi
Names registered in the model: con, x, θ

In [27]:
objective_value(firstst)

-126.0

In [28]:
value.(x)

2-element Vector{Float64}:
 40.0
 30.0

The solution is therefore $x = (40,\ 30)$.

We can do the same in the extended form

In [29]:
m = Model()
scenarios = [X; Y]
prob = 1/100 # probability does not matter here (assuming i.i.d. variables)

@variable(m, x[1:nBreadTypes] >= 0)
@variable(m, y[1:nScenarios, 1:nBreadTypes] >= 0)
@variable(m, w[1:nScenarios, 1:nBreadTypes] >= 0)

@constraint(m, sum(productionFlour[i]*x[i] for i in 1:nBreadTypes) <= maxFlour)

for j = 1:nBreadTypes
    @constraint(m, x[j] in MOI.Integer())
    for s = 1:nScenarios
        @constraint(m, y[s,j] <= scenarios[j, s])
        @constraint(m, y[s,j] + w[s,j] <= x[j])
        @constraint(m, w[s,j] >= x[j]-scenarios[j, s])
        @constraint(m, y[s,j] <= x[j])
    end
end

@objective(m, Min, sum(productionCosts[i]*x[i] for i in 1:nBreadTypes) -
    sum(prob*sum((q[j]*y[s, j] + r[j]*w[s, j]) for j = 1:nBreadTypes) for s=1:nScenarios)
)

println(m)

Min 1.5 x[1] + 1.8 x[2] - 0.03 y[1,1] - 0.01 w[1,1] - 0.04 y[1,2] - 0.012 w[1,2] - 0.03 y[2,1] - 0.01 w[2,1] - 0.04 y[2,2] - 0.012 w[2,2] - 0.03 y[3,1] - 0.01 w[3,1] - 0.04 y[3,2] - 0.012 w[3,2] - 0.03 y[4,1] - 0.01 w[4,1] - 0.04 y[4,2] - 0.012 w[4,2] - 0.03 y[5,1] - 0.01 w[5,1] - 0.04 y[5,2] - 0.012 w[5,2] - 0.03 y[6,1] - 0.01 w[6,1] - 0.04 y[6,2] - 0.012 w[6,2] - 0.03 y[7,1] - 0.01 w[7,1] - 0.04 y[7,2] - 0.012 w[7,2] - 0.03 y[8,1] - 0.01 w[8,1] - 0.04 y[8,2] - 0.012 w[8,2] - 0.03 y[9,1] - 0.01 w[9,1] - 0.04 y[9,2] - 0.012 w[9,2] - 0.03 y[10,1] - 0.01 w[10,1] - 0.04 y[10,2] - 0.012 w[10,2] - 0.03 y[11,1] - 0.01 w[11,1] - 0.04 y[11,2] - 0.012 w[11,2] - 0.03 y[12,1] - 0.01 w[12,1] - 0.04 y[12,2] - 0.012 w[12,2] - 0.03 y[13,1] - 0.01 w[13,1] - 0.04 y[13,2] - 0.012 w[13,2] - 0.03 y[14,1] - 0.01 w[14,1] - 0.04 y[14,2] - 0.012 w[14,2] - 0.03 y[15,1] - 0.01 w[15,1] - 0.04 y[15,2] - 0.012 w[15,2] - 0.03 y[16,1] - 0.01 w[16,1] - 0.04 y[16,2] - 0.012 w[16,2] - 0.03 y[17,1] - 0.01 w[17,1] - 0.04

 -x[2] + w[4,2] ≥ -33.23571242567453
 -x[2] + w[5,2] ≥ -27.05481719217093
 -x[2] + w[6,2] ≥ -31.546719726200628
 -x[2] + w[7,2] ≥ -31.868800355840406
 -x[2] + w[8,2] ≥ -28.61120169613001
 -x[2] + w[9,2] ≥ -27.006790364878132
 -x[2] + w[10,2] ≥ -28.47662381456601
 -x[2] + w[11,2] ≥ -30.203909392472553
 -x[2] + w[12,2] ≥ -30.417291630637497
 -x[2] + w[13,2] ≥ -29.482820827357322
 -x[2] + w[14,2] ≥ -29.146178113770233
 -x[2] + w[15,2] ≥ -26.506540502438888
 -x[2] + w[16,2] ≥ -31.3316286453197
 -x[2] + w[17,2] ≥ -30.77227199389245
 -x[2] + w[18,2] ≥ -30.3624695636054
 -x[2] + w[19,2] ≥ -31.970163708698166
 -x[2] + w[20,2] ≥ -30.690800318424742
 -x[2] + w[21,2] ≥ -28.39089557663169
 -x[2] + w[22,2] ≥ -29.62276871195132
 -x[2] + w[23,2] ≥ -32.948050168529626
 -x[2] + w[24,2] ≥ -29.97427289773172
 -x[2] + w[25,2] ≥ -30.070497429533567
 -x[2] + w[26,2] ≥ -32.33543575956018
 -x[2] + w[27,2] ≥ -30.155914965094137
 -x[2] + w[28,2] ≥ -33.141664856968745
 -x[2] + w[29,2] ≥ -29.01960659623389
 -x[2]

 -x[1] + y[70,1] + w[70,1] ≤ 0.0
 -x[1] + y[70,1] ≤ 0.0
 y[71,1] ≤ 44.832839779794696
 -x[1] + y[71,1] + w[71,1] ≤ 0.0
 -x[1] + y[71,1] ≤ 0.0
 y[72,1] ≤ 48.47769656730959
 -x[1] + y[72,1] + w[72,1] ≤ 0.0
 -x[1] + y[72,1] ≤ 0.0
 y[73,1] ≤ 54.24485998809533
 -x[1] + y[73,1] + w[73,1] ≤ 0.0
 -x[1] + y[73,1] ≤ 0.0
 y[74,1] ≤ 35.42246970030651
 -x[1] + y[74,1] + w[74,1] ≤ 0.0
 -x[1] + y[74,1] ≤ 0.0
 y[75,1] ≤ 49.9236070604481
 -x[1] + y[75,1] + w[75,1] ≤ 0.0
 -x[1] + y[75,1] ≤ 0.0
 y[76,1] ≤ 46.66805418844045
 -x[1] + y[76,1] + w[76,1] ≤ 0.0
 -x[1] + y[76,1] ≤ 0.0
 y[77,1] ≤ 63.03970344001145
 -x[1] + y[77,1] + w[77,1] ≤ 0.0
 -x[1] + y[77,1] ≤ 0.0
 y[78,1] ≤ 57.10704158206708
 -x[1] + y[78,1] + w[78,1] ≤ 0.0
 -x[1] + y[78,1] ≤ 0.0
 y[79,1] ≤ 51.589407079720786
 -x[1] + y[79,1] + w[79,1] ≤ 0.0
 -x[1] + y[79,1] ≤ 0.0
 y[80,1] ≤ 56.424644667924554
 -x[1] + y[80,1] + w[80,1] ≤ 0.0
 -x[1] + y[80,1] ≤ 0.0
 y[81,1] ≤ 43.09467084050766
 -x[1] + y[81,1] + w[81,1] ≤ 0.0
 -x[1] + y[81,1] ≤ 0.0
 y[82,1

 y[83,2] ≤ 29.107767443983764
 -x[2] + y[83,2] + w[83,2] ≤ 0.0
 -x[2] + y[83,2] ≤ 0.0
 y[84,2] ≤ 29.669356299317506
 -x[2] + y[84,2] + w[84,2] ≤ 0.0
 -x[2] + y[84,2] ≤ 0.0
 y[85,2] ≤ 28.855794810508556
 -x[2] + y[85,2] + w[85,2] ≤ 0.0
 -x[2] + y[85,2] ≤ 0.0
 y[86,2] ≤ 31.524153694687314
 -x[2] + y[86,2] + w[86,2] ≤ 0.0
 -x[2] + y[86,2] ≤ 0.0
 y[87,2] ≤ 29.025421355559875
 -x[2] + y[87,2] + w[87,2] ≤ 0.0
 -x[2] + y[87,2] ≤ 0.0
 y[88,2] ≤ 27.89402503493531
 -x[2] + y[88,2] + w[88,2] ≤ 0.0
 -x[2] + y[88,2] ≤ 0.0
 y[89,2] ≤ 31.763317150083743
 -x[2] + y[89,2] + w[89,2] ≤ 0.0
 -x[2] + y[89,2] ≤ 0.0
 y[90,2] ≤ 30.064487527647778
 -x[2] + y[90,2] + w[90,2] ≤ 0.0
 -x[2] + y[90,2] ≤ 0.0
 y[91,2] ≤ 34.72462054751108
 -x[2] + y[91,2] + w[91,2] ≤ 0.0
 -x[2] + y[91,2] ≤ 0.0
 y[92,2] ≤ 31.922742479279666
 -x[2] + y[92,2] + w[92,2] ≤ 0.0
 -x[2] + y[92,2] ≤ 0.0
 y[93,2] ≤ 34.031874555233394
 -x[2] + y[93,2] + w[93,2] ≤ 0.0
 -x[2] + y[93,2] ≤ 0.0
 y[94,2] ≤ 29.327477653948115
 -x[2] + y[94,2] + w[94,2]

In [30]:
set_optimizer(m, solver)
optimize!(m)

Academic license - for non-commercial use only - expires 2022-10-15
Gurobi Optimizer version 9.1.2 build v9.1.2rc0 (mac64)
Thread count: 4 physical cores, 8 logical processors, using up to 8 threads
Optimize a model with 801 rows, 402 columns and 1602 nonzeros
Model fingerprint: 0xc2afa4f5
Variable types: 400 continuous, 2 integer (0 binary)
Coefficient statistics:
  Matrix range     [1e+00, 2e+02]
  Objective range  [1e-02, 2e+00]
  Bounds range     [0e+00, 0e+00]
  RHS range        [3e+01, 1e+04]
Found heuristic solution: objective 0.0000000
Presolve removed 798 rows and 198 columns
Presolve time: 0.00s
Presolved: 3 rows, 204 columns, 206 nonzeros
Variable types: 202 continuous, 2 integer (0 binary)

Root relaxation: objective -1.246684e+02, 4 iterations, 0.00 seconds

    Nodes    |    Current Node    |     Objective Bounds      |     Work
 Expl Unexpl |  Obj  Depth IntInf | Incumbent    BestBd   Gap | It/Node Time

     0     0 -124.66843    0    2    0.00000 -124.66843      -     

In [31]:
value.(x)

2-element Vector{Float64}:
 44.0
 27.0

In [32]:
objective_value(m)

-124.56321545868644

The solution is therefore $x = (44,\ 27)$.